### PR TITLE
Allow for negative numbers to be parsed

### PR DESCRIPTION
**Pull request by cccs-rs**
YARA recognizes having negative numerical values, but plyara will crash if it encounters a HYPHEN with a NUM type when parsing a rule.

Related external issue: https://github.com/CybercentreCanada/assemblyline/issues/231

### CODE REVIEW

**Comment by utkonos** [comment]
@cccs-rs Thanks for all of this!

I cherry-pick merged almost all of these changes. I needed to decouple the keyword changes and the black formatting changes. Also, it looks like in that commit you removed two keywords that are still part of yara by accident. I added them back plus the one you added. I also brought the keywords in line with the current documentation. I also added all the missing modules that exist in the source code only line dex and macho as well as lnk which only exists in yara-x. This project uses single quotes rather than double. There may be a black setting for that? I like some of the other black formatting and adopted those. I (and the linters) disagree with black on the other formatting changes. Finally, those docstring looking single quotes in `core.py` are deliberate. They're not docstrings. They are a quirk of PLY. I had to adjust my linters to accomodate them. One of the projects that I'd like to complete is the transition to SLY, dabeaz's updated project that replaces PLY. I took his compilers course during the pandemic and learned how to use it. I started on that before I had to stop working on plyara for a bit.

If you're amenable, I'd like to add you as a maintainer for this project so we can work on it together.

I will be reworking the unit tests so that they're more organized over the weekend and will be cutting a 2.2 on pypi when that's done and they're passing.

**Comment by cccs-rs** [comment]
Sure thing! I was thinking of writing a YARA parser in Lark because I noticed there was some issues with processing keywords/functions from imports like PE (and I also didn't know if this project was being actively maintained 😅).

Sample code for some basic parsing that I wrote as a hobby:
```python
yara_parser = Lark(r"""
    start: yara_rule*
    yara_rule       : rule_imports [rule_scope] "rule" rule_name [":" rule_tags] "{" ["meta:" meta] ["strings:" strings] "condition:" condition_terms "}"
    rule_name       :   /\w+/
    rule_scope      :   /(private|global)/

    rule_tags       : rule_tag*
    rule_tag        : /\w+/

    rule_imports    :   rule_import*
    rule_import     :   "import" "\"" /(pe|elf|lnk|dotnet|macho|console|hash|math|string|time)/ "\""

    // META
    meta        : meta_pair*
    meta_pair   : meta_key "=" meta_value
    meta_key    : /[a-z0-9_]+/
    meta_value  : ESCAPED_STRING | SIGNED_NUMBER | /(true|false)/

    // STRINGS
    strings         : string_pair*
    string_pair     : string_key "=" string_value
    string_key      : /\$[a-z0-9_]+/
    string_value    : text_pattern | hex_pattern | regex_pattern

    // STRINGS: HEX
    hex_pattern     : "{" [byte|jump]* [hex_regex]* [byte|jump]* "}" hex_modifiers
    hex_regex       : "(" [byte|jump]* [/\|/ [byte|jump]*]* ")"
    hex_modifiers   : [hex_modifier]*
    hex_modifier    : /(private)/
    byte            : /(~)?[a-fA-F0-9?]{2}/
    jump            : "[" [SIGNED_NUMBER] "-" [SIGNED_NUMBER] "]" | "[" SIGNED_NUMBER "]"

    // STRINGS: TEXT
    text_pattern    : ESCAPED_STRING text_modifiers
    text_modifiers  : [text_modifier]*
    text_modifier   : /(xor(\(0x[0-9a-fA-F]{2}-0x[0-9a-fA-F]{2}\))?|base64(wide)?(\(".+"\))?|nocase|wide|ascii|fullword|private)/

    // STRINGS: REGEX
    regex_pattern:   /\/(.+)\/(i)?(s)?/ regex_modifiers
    regex_modifiers : [regex_modifier]*
    regex_modifier  : /(nocase|wide|ascii|fullword|private)/

    // CONDITIONS
    condition_terms : condition_term*
    condition_term : /\w+/ | string_key

    %import common.ESCAPED_STRING
    %import common.SIGNED_NUMBER
    %import common.HEXDIGIT
    %import common.WS
    %ignore WS
""")
```

But working on this together would be a great opportunity for collaboration! I haven't used SLY before but it's not like I'm an expert in writing parsers anyways so it would be neat to learn😁

**Comment by utkonos** [comment]
I've added another fellow threat intel traveller to the project as a maintainer. I'll send you an invite as well. No need to keep different forks, and I would love to collaborate.

**Comment by utkonos** [comment]
@larsborn @cccs-rs 
Lets plan of a little get together via whatever kind of chat or comms you all like so we can do some coordination. Zoom works or if you have a jitsi instance or whatever works for you two

**Comment by utkonos** [comment]
I enabled discussions on the repo. We can move this all here:

https://github.com/orgs/plyara/discussions/140

File: plyara/core.py [hunk 1]
```diff
@@ -293,210 +351,223 @@ def parse_string(self, input_string):
 
         for rule in self.rules:
             if any(self.imports):
-                rule['imports'] = list(self.imports)
+                rule["imports"] = [yara_import for yara_import in self.imports if f"{yara_import}." in rule['raw_condition'] ]
```
Comment (utkonos) [inline plyara/core.py:354]: This change introduces a bug. If a user has set `store_raw_sections` to `False`, then `raw_condition` does not exist. The check for import usage needs to be performed against the condition list.

I am fixing it.

https://github.com/plyara/plyara/issues/141

**Pull request closed without merging**